### PR TITLE
(43909856) ViT (Visual Transformer) for image classification of ADNI dataset

### PR DESCRIPTION
**Pull request by retinotopyproject**
Hi there,

This fork of the 'topic-recognition' branch contains code for a ViT (a Visual Transformer) model. This is based on the ViT-S/16
model, from the IEEE conference paper ["Scaling Vision Transformers" (Zhai et al., 2022).](https://ieeexplore.ieee.org/document/9880094)

This model was used to perform binary classification of images from the Alzheimer's Disease Neuroimaging Initiative (ADNI) MRI dataset. The ViT attempts to classify 2D MRI slices as belonging to either Cognitive Normal (NC) or Alzheimer's Detected (AD) patients.

For more detailed information about the model, its performance, and any dependencies/requirements for training or testing it, please see the README file. All files related to the ViT model, including the README, are in the directory location _'recognition/TRANSFORMER_43909856'_.

The files added include:

- **dataset.py:** Loads the ADNI dataset images and their corresponding image labels, and performs the required train/validation/test set splits. Also applies any required pre-processing or transforms to the data. A basic plot of the input images (after pre-processing transforms) can be generated from a method in this file.

- **modules.py:** Contains the modules required to create the ViT model network and its sub-components.

- **train.py:** Trains the model on the ADNI dataset over the specified number of epochs. Evaluates the model on the validation set during the training process, at the end of every epoch. Also contains methods for plotting training and validation loss (and validation accuracy) throughout training.

- **predict.py:** Used to run inference (make predictions of classes) on the model. Also contains a method for generating a confusion matrix, based on the test set's predicted classes and empirical/observed classes.

- Additional plots can be found in the _'plots'_ subdirectory.

### CODE REVIEW

**Comment by LinfengLiu98** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme:
Excellent

Commit messages: Excellent, detailed.

Code:
- Uses Vision transformer
- Well commented
- Good model design

Functionality/Performance:
- Test accuracy is 60.37%
- Did patient-level split
- Preprocssing data correctly
- Performed data augmentation
- Well done

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable/good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good

**Comment by shakes76** [comment]
Please remove gitignore file for merge, does not affect grade only merge.

**Comment by wangzhaomxy** [comment]
No feedback attempt and no feedback marks lost.

**Comment by retinotopyproject** [comment]
Hi @shakes76, I have resolved the merge conflict related to the .gitignore file. Thanks

**Comment by wangzhaomxy** [comment]
Feedback attempt and no feedback marks lost.
